### PR TITLE
Implement filteradd

### PR DESCRIPTION
**Pull request by supreme2580**
Closes: https://github.com/zig-bitcoin/btczee/issues/93

### CODE REVIEW

File: src/network/protocol/messages/filteradd.zig [hunk 1]
```diff
@@ -0,0 +1,78 @@
+const std = @import("std");
+const protocol = @import("../lib.zig");
+
+/// FilterAddMessage represents the "filteradd" message
+///
+/// https://developer.bitcoin.org/reference/p2p_networking.html#filteradd
+pub const FilterAddMessage = struct {
+    element: []const u8,
```
Comment (tdelabro) [inline src/network/protocol/messages/filteradd.zig:8]: you need a deinit method to free this

File: src/network/protocol/messages/filteradd.zig [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        return ret;
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+    pub fn serializeToSlice(self: *const Self, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        try self.serializeToWriter(fbs.writer());
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Self {
```
Comment (tdelabro) [inline src/network/protocol/messages/filteradd.zig:44]: comptime {
            if (!std.meta.hasFn(@TypeOf(w), "readAllAlloc")) @compileError("Expects r to have fn 'readAllAlloc'.");

File: src/network/protocol/messages/filteradd.zig [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        const element = try r.readAllAlloc(allocator, 520);
+        errdefer allocator.free(element);
```
Comment (tdelabro) [inline src/network/protocol/messages/filteradd.zig:46]: there is no way if fails before returning. Remove this line

File: src/network/protocol/messages/filteradd.zig [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    pub fn checksum(self: *const Self) [4]u8 {
+        var digest: [32]u8 = undefined;
+        std.crypto.hash.sha2.Sha256.hash(self.element, &digest, .{});
+        std.crypto.hash.sha2.Sha256.hash(&digest, &digest, .{});
+        return digest[0..4].*;
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    pub fn serializeToWriter(self: *const Self, w: anytype) !void {
+        try w.writeAll(self.element);
```
Comment (tdelabro) [inline src/network/protocol/messages/filteradd.zig:26]: comptime {
        if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects r to have fn 'writeAll'.");

File: src/network/protocol/messages/lib.zig [hunk 5]
```diff
@@ -62,6 +66,7 @@ pub const Message = union(MessageTypes) {
             .sendcmpct => {},
             .feefilter => {},
             .filterclear => {},
+            .filteradd => {},
```
Comment (tdelabro) [inline src/network/protocol/messages/lib.zig:69]: and use deinit here

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/network/protocol/messages/lib.zig (hunk 5 above)
Comment (supreme2580) [reply src/network/protocol/messages/lib.zig:69]: Okay boss going over them all, will update it asap

File: src/network/protocol/messages/filteradd.zig (hunk 1 above)
Comment (supreme2580) [reply src/network/protocol/messages/filteradd.zig:8]: @tdelabro I have a deinit method on line 59-61

`
    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
        allocator.free(self.element);
    }
`

File: src/network/protocol/messages/filteradd.zig (hunk 3 above)
Comment (supreme2580) [reply src/network/protocol/messages/filteradd.zig:46]: removed it

File: src/network/protocol/messages/filteradd.zig (hunk 4 above)
Comment (supreme2580) [reply src/network/protocol/messages/filteradd.zig:26]: added this above `try w.writeAll(self.element);`

**Review by tdelabro — APPROVED** [review]

**Pull request merged**